### PR TITLE
✨ Supporting externally managed Control Plane

### LINKED ISSUE #4437: Supporting externally managed Control Plane
/kind feature

**Describe the solution you'd like**

[Kamaji](https://github.com/clastix/kamaji) is an Open Source project that is offering Kubernetes Control Plane as a Service. Thanks to the CAPI Control Plane provider contract and its [implementation](https://github.com/clastix/cluster-api-control-plane-provider-kamaji) it allows creating workload clusters just made of worker nodes and consuming the kube-apiserver as an external service.

At the current state, the Cluster API Provider for AWS is not supporting this contract since it gives for granted the Control Plane is made of Virtual Machines on AWS, thus provisioning of an AWS Load Balancer, or the reference to an already created one is mandatory and doesn't provide a chance to support an externally managed.

As briefly discussed with @richardcase, an annotation for the `AWSCluster` could skip the creation of the AWS Load Balancer: in a such situation, the AWS Cluster API provider will wait for the Control Plane provider to patch the `/AWSCluster/spec/controlPlaneEndpoint` object with the externally managed endpoint.

**Anything else you would like to add:**

The proposed implementation sounds similar to what we achieved, as Kamaji developers, with other infrastructure providers we tested, such as vSphere, KubeVirt, Equinix, and OpenStack (although the latter supports a field to reference a fixed API server address and port).

I'm already working on a PR to address this use case and discussing the implementation.

**Environment:**

- Cluster-api-provider-aws version: v2.2.1
- Kubernetes version: (use `kubectl version`): N.R.
- OS (e.g. from `/etc/os-release`): N.R.

### PR DESCRIPTION
**Pull request by prometherion**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Cluster API allows defining externally managed control plane for the given cluster, the current code-base doesn't consider this since it nevertheless performs the following actions:

- reconciliation of a Load Balancer, although already provisioned by the ControlPlane provider
- blocking the EC2 creation since it requires an ELB DNS name

**Which issue(s) this PR fixes**:
Fixes #4437

**Special notes for your reviewer**:

**Checklist**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests
- [x] adds or updates e2e tests

**Release note**:

```release-note
A new enum value, `disabled`, is added for the `AWSCluster.spec.controlPlaneLoadBalancer.loadBalancerType` field, which skips the reconciliation of the load balancer for the given cluster, useful for clusters which are consuming an externally managed Control Plane.
```

### CODE REVIEW

File: pkg/cloud/scope/machine.go [hunk 1]
```diff
@@ -370,6 +372,41 @@ func (m *MachineScope) IsEKSManaged() bool {
 	return m.InfraCluster.InfraCluster().GetObjectKind().GroupVersionKind().Kind == ekscontrolplanev1.AWSManagedControlPlaneKind
 }
 
+func (m *MachineScope) IsControlPlaneExternallyManaged() bool {
```
Comment (prometherion) [inline pkg/cloud/scope/machine.go:375]: Honestly, I'm not so much happy about this solution, since several steps are returning errors, and we should avoid that at the beginning for the scope.

I'd suggest introducing a new function to the interface `ClusterScoper`,  such as `ControlPlane`, which returns the Control Plane object in order to check its status if it is externally managed.

Comment (prometherion) [reply pkg/cloud/scope/machine.go:380]: Addressed with changes from be4b28aff.

File: controllers/awscluster_controller.go [hunk 2]
```diff
@@ -252,6 +262,44 @@ func (r *AWSClusterReconciler) reconcileDelete(ctx context.Context, clusterScope
...
+		return nil, err
+	}
+
+	if awsCluster.Status.Network.APIServerELB.DNSName == "" {
+		conditions.MarkFalse(awsCluster, infrav1.LoadBalancerReadyCondition, infrav1.WaitForDNSNameReason, clusterv1.ConditionSeverityInfo, "")
+		clusterScope.Info("Waiting on API server ELB DNS name")
+		return pointer.Duration(15 * time.Second), nil
+	}
+
+	clusterScope.Debug("looking up IP address for DNS", "dns", awsCluster.Status.Network.APIServerELB.DNSName)
```
Comment (vincepri) [inline controllers/awscluster_controller.go:286]:
```suggestion
	clusterScope.Debug("Looking up IP address for DNS", "dns", awsCluster.Status.Network.APIServerELB.DNSName)
```

File: controllers/awsmachine_controller.go [hunk 3]
```diff
@@ -1141,3 +1149,19 @@ func (r *AWSMachineReconciler) ensureInstanceMetadataOptions(ec2svc services.EC2
 
 	return ec2svc.ModifyInstanceMetadataOptions(instance.ID, machine.Spec.InstanceMetadataOptions)
 }
+
+func (r *AWSMachineReconciler) getControlPlane(ctx context.Context, log *logger.Logger, cluster *clusterv1.Cluster) (*unstructured.Unstructured, error) {
+	controlPlane := unstructured.Unstructured{}
```
Comment (vincepri) [inline controllers/awsmachine_controller.go:1154]:
```suggestion
	controlPlane := new(unstructured.Unstructured)
```
So we carry the pointer around?

File: controllers/awscluster_controller.go (hunk 2 above)
Comment (prometherion) [reply controllers/awscluster_controller.go:286]: That message was inherited from the previous changes:  https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/2891a57069b8365da6c601c2aa8a9ea9ad7cde7c/controllers/awscluster_controller.go#L316

We could address this renaming in a different PR to avoid too many files changes.

**Comment by prometherion** [comment]
If the proposed approach is ok I can work on proper unit-tests for this.

File: controllers/awsmachine_controller.go (hunk 3 above)
Comment (muraee) [reply controllers/awsmachine_controller.go:1154]: I think you can use `Get()` func from [sigs.k8s.io/cluster-api/controllers/external](https://github.com/kubernetes-sigs/cluster-api/blob/3a373971a953cdd1db97cdba89db5fba00452041/controllers/external/util.go#L34) pkg here

Comment (prometherion) [reply controllers/awsmachine_controller.go]: Thanks, @muraee, implemented with bc6370bdb.

File: controllers/awscluster_controller.go [hunk 4]
```diff
@@ -252,6 +262,44 @@ func (r *AWSClusterReconciler) reconcileDelete(ctx context.Context, clusterScope
 	return reconcile.Result{}, nil
 }
 
+func (r *AWSClusterReconciler) reconcileLoadBalancer(clusterScope *scope.ClusterScope, awsCluster *infrav1.AWSCluster) (*time.Duration, error) {
+	if annotations.Has(awsCluster, DisableLoadBalancerReconciliationAnnotation) {
+		clusterScope.Info("load balancer reconciliation disabled")
```
[the remark this answers is not in the document]
Comment (Skarlso) [reply controllers/awscluster_controller.go:267]: Yep. Seconded the Debug level.

[the remark this answers is not in the document]
Comment (Skarlso) [reply controllers/awscluster_controller.go:267]: Actually... I would like to see in the log. But slightly different wording. Reconciliation is _not_ disabled. We still reconcile, we are just checking for external load balancer. So say something like this instead:

`load balancer reconciliation shifted to external provider, checking external endpoint`.

File: controllers/awsmachine_controller.go [hunk 5]
```diff
@@ -197,10 +199,16 @@ func (r *AWSMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 
 	infrav1.SetDefaults_AWSMachineSpec(&awsMachine.Spec)
 
+	cp, err := r.getControlPlane(ctx, log, cluster)
```
Comment (Skarlso) [inline controllers/awsmachine_controller.go:202]: I'm sorry, I'm probably missing something, but how is this related change? 🤔

File: controllers/awscluster_controller.go (hunk 2 above)
Comment (Ankitasw) [reply controllers/awscluster_controller.go:286]: Since its a minor nit change, feel free to do it in this PR itself

File: controllers/awscluster_controller.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (prometherion) [reply controllers/awscluster_controller.go]: Addressed with b438e14f2.

File: controllers/awsmachine_controller.go (hunk 5 above)
Comment (prometherion) [reply controllers/awsmachine_controller.go:203]: The Control Plane object is required since the Cluster API contract to determine if it's externally managed resides on its status (`/status/externalManagedControlPlane` boolean value).

Without it, we cannot understand if the scope is actually `IsControlPlaneExternallyManaged` or not.

Comment (Skarlso) [reply controllers/awsmachine_controller.go:203]: Understood. Thanks for the explanation!

File: controllers/awscluster_controller.go (hunk 2 above)
Comment (prometherion) [reply controllers/awscluster_controller.go]: Addressed with 58e62e7b0

File: controllers/awscluster_controller.go [hunk 6]
```diff
@@ -58,6 +60,14 @@ import (
 	"sigs.k8s.io/cluster-api/util/predicates"
 )
 
+const (
+	// DisableLoadBalancerReconciliationAnnotation allows skipping the reconciliation
+	// of a Load Balancer for the given AWSCluster.
+	// This annotation is used then a Control Plane provider, such as an external one,
+	// is providing by itself the controlPlaneEndpoint resource.
+	DisableLoadBalancerReconciliationAnnotation = "sigs.k8s.io/cluster-api-provider-aws-disable-loadbalancer"
```
[the remark this answers is not in the document]
Comment (vincepri) [reply controllers/awscluster_controller.go:69]: A new field is a backward compatible change and does not require a new API revision

File: controllers/awscluster_controller.go [hunk 7]
```diff
@@ -252,6 +263,44 @@ func (r *AWSClusterReconciler) reconcileDelete(ctx context.Context, clusterScope
 	return nil
 }
 
+func (r *AWSClusterReconciler) reconcileLoadBalancer(clusterScope *scope.ClusterScope, awsCluster *infrav1.AWSCluster) (*time.Duration, error) {
+	if annotations.Has(awsCluster, DisableLoadBalancerReconciliationAnnotation) {
+		clusterScope.Debug("load balancer reconciliation shifted to external provider, checking external endpoint")
+
+		return r.checkForExternalControlPlaneLoadBalancer(clusterScope, awsCluster), nil
+	}
+
```
[the remark this answers is not in the document]
Comment (prometherion) [reply controllers/awscluster_controller.go:272]: We don't know if the externally managed control plane will create its own IP address, as e.g. Kamaji is going, or rather because the user is hard-coding a pre-defined IP address in the `/AWSCluster/spec/controlPlaneEndpoint` field.

In such edge-case scenarios, the knob through the annotation will skip the creation of the infrastructure components that would be unnecessary.

If you prefer, we could rework the PR and embed this logic in the retrieval of the Control Plane object: if it's marked as externally managed, we can skip the CP infrastructure components.

[the remark this answers is not in the document]
Comment (prometherion) [reply controllers/awscluster_controller.go]: Fixed by adding a field to the `AWSCluster` type, which made this outdated.

File: controllers/awscluster_controller.go [hunk 8]
```diff
@@ -442,3 +460,27 @@ func (r *AWSClusterReconciler) requeueAWSClusterForUnpausedCluster(_ context.Con
...
 	}
 }
+
+func (r *AWSClusterReconciler) checkForExternalControlPlaneLoadBalancer(clusterScope *scope.ClusterScope, awsCluster *infrav1.AWSCluster) *time.Duration {
+	switch {
+	case len(awsCluster.Spec.ControlPlaneEndpoint.Host) == 0 && awsCluster.Spec.ControlPlaneEndpoint.Port == 0:
+		clusterScope.Info("AWSCluster control plane endpoint is still non-populated")
+		conditions.MarkFalse(awsCluster, infrav1.LoadBalancerReadyCondition, infrav1.WaitForExternalControlPlaneEndpointReason, clusterv1.ConditionSeverityInfo, "")
+
+		return pointer.Duration(15 * time.Second)
```
Comment (faiq) [inline controllers/awscluster_controller.go:470]: nit: let's make the duration as a constant

File: api/v1beta2/awscluster_types.go [hunk 9]
```diff
@@ -221,7 +222,7 @@ type AWSLoadBalancerSpec struct {
 
 	// LoadBalancerType sets the type for a load balancer. The default type is classic.
 	// +kubebuilder:default=classic
-	// +kubebuilder:validation:Enum:=classic;elb;alb;nlb
+	// +kubebuilder:validation:Enum:=classic;elb;alb;nlb;disabled
```
Comment (faiq) [inline api/v1beta2/awscluster_types.go:225]: One thing that I think would be helpful is having a webhook validation that looks to see if the other related LoadBalancer options are not set. 

For example, I shouldn't be able to set things like `Scheme, HealthCheckProtocol, AdditionalSecurityGroups` and so on if the loadbalancertype == disabled

**Comment by dlipovetsky** [comment]
In today's CAPA office hours, we talked about adding a test that will verify the expected behavior, and prevent CAPA from breaking downstream implementations of a "managed control plane." During the discussion, an end-to-end test did not seem practical, since it requires running one (or more) of these implementations, but @prometherion will explore some other approaches.

File: api/v1beta2/awscluster_types.go (hunk 9 above)
Comment (prometherion) [reply api/v1beta2/awscluster_types.go:225]: Implemented.

**Comment by prometherion** [comment]
I think this is ready, I tested it personally by combining it with Kamaji.

```yaml
apiVersion: cluster.x-k8s.io/v1beta1
kind: Cluster
metadata:
  labels:
    ccm: external
    csi: external
  name: capi-quickstart
  namespace: default
spec:
  clusterNetwork:
    pods:
      cidrBlocks:
      - 192.168.0.0/16
  controlPlaneRef:
    apiVersion: controlplane.cluster.x-k8s.io/v1beta1
    kind: KamajiControlPlane
    name: capi-quickstart-control-plane
  infrastructureRef:
    apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
    kind: AWSCluster
    name: capi-quickstart
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AWSCluster
metadata:
  name: capi-quickstart
  namespace: default
spec:
  region: us-east-1
  sshKeyName: default
  controlPlaneLoadBalancer:
    loadBalancerType: disabled
---
apiVersion: controlplane.cluster.x-k8s.io/v1alpha1
kind: KamajiControlPlane
metadata:
  name: capi-quickstart-control-plane
  namespace: default
spec:
  dataStoreName: default
  addons:
    coreDNS: { }
    kubeProxy: { }
  kubelet:
    cgroupfs: systemd
    preferredAddressTypes:
      - ExternalIP
      - InternalIP
      - Hostname
  network:
    serviceAddress: 78.134.89.204
    serviceType: ClusterIP
  deployment:
  replicas: 2
  version: 1.26.0
---
apiVersion: cluster.x-k8s.io/v1beta1
kind: MachineDeployment
metadata:
  name: capi-quickstart-md-0
  namespace: default
spec:
  clusterName: capi-quickstart
  replicas: 2
  selector:
    matchLabels: null
  template:
    spec:
      bootstrap:
        configRef:
          apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
          kind: KubeadmConfigTemplate
          name: capi-quickstart-md-0
      clusterName: capi-quickstart
      infrastructureRef:
        apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
        kind: AWSMachineTemplate
        name: capi-quickstart-md-0
      version: v1.26.0
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta2
kind: AWSMachineTemplate
metadata:
  name: capi-quickstart-md-0
  namespace: default
spec:
  template:
    spec:
      iamInstanceProfile: nodes.cluster-api-provider-aws.sigs.k8s.io
      instanceType: t3.large
---
apiVersion: bootstrap.cluster.x-k8s.io/v1beta1
kind: KubeadmConfigTemplate
metadata:
  name: capi-quickstart-md-0
  namespace: default
spec:
  template:
    spec:
      joinConfiguration:
        nodeRegistration:
          kubeletExtraArgs:
            cloud-provider: external
          name: '{{ ds.meta_data.local_hostname }}'
```

> JFI, the Control Plane is running on my lab environment, outside of AWS

Once the provisioning is complete, the nodes are joining as expected:
```
NAME                          STATUS     ROLES    AGE    VERSION
ip-10-0-114-74.ec2.internal   NotReady   <none>   112s   v1.26.0
ip-10-0-96-155.ec2.internal   NotReady   <none>   113s   v1.26.0
```

The `clusterctl` describe output:
```
NAME                                                               READY  SEVERITY  REASON                       SINCE  MESSAGE                                                                
Cluster/capi-quickstart                                            True                                          3m20s                                                                          
├─ClusterInfrastructure - AWSCluster/capi-quickstart               True                                          4m6s                                                                           
├─ControlPlane - KamajiControlPlane/capi-quickstart-control-plane                                                                                                                               
└─Workers                                                                                                                                                                                       
  └─MachineDeployment/capi-quickstart-md-0                         False  Warning   WaitingForAvailableMachines  8m22s  Minimum availability requires 2 replicas, current 0 available           
    └─2 Machines...                                                True                                          2m32s  See capi-quickstart-md-0-lp5xc-f4bbx, capi-quickstart-md-0-lp5xc-q254x  
```

> please, notice I skipped the creation of the CCM along with the CNI plugin for the sake of speed testing

Could we trigger again the e2e testing, please?

**Comment by richardcase** [comment]
Until #4733 merges

/hold

But this looks good to me, pending any changes required after 4733.

**Comment by prometherion** [comment]
Thanks, @richardcase, rebased as requested.

The failed test seems unrelated to the changes we introduced.

**Comment by richardcase** [comment]
Thanks @prometherion 

/unhold

**Comment by faiq** [comment]
the tests are all passing, and the code looks good. i think this is ready to merge

File: api/v1beta2/awscluster_webhook.go [hunk 10]
```diff
@@ -298,5 +298,49 @@ func (r *AWSCluster) validateControlPlaneLBs() field.ErrorList {
 		}
 	}
 
+	if r.Spec.ControlPlaneLoadBalancer.LoadBalancerType == LoadBalancerTypeDisabled {
+		if r.Spec.ControlPlaneLoadBalancer.Name != nil {
```
Comment (richardcase) [inline api/v1beta2/awscluster_webhook.go:302]: I'm wondering if we need a `func Empty() bool` on the load balancer instead of checking each sub field. But lets discuss that separately.

File: controllers/awsmachine_controller.go [hunk 11]
```diff
@@ -1178,3 +1186,22 @@ func (r *AWSMachineReconciler) ensureInstanceMetadataOptions(ec2svc services.EC2
 
 	return ec2svc.ModifyInstanceMetadataOptions(instance.ID, machine.Spec.InstanceMetadataOptions)
 }
+
+// +kubebuilder:rbac:groups=controlplane.cluster.x-k8s.io,resources=*,verbs=get;list;watch
```
[the remark this answers is not in the document]
Comment (prometherion) [reply controllers/awsmachine_controller.go:1209]: TBH, this wildcard makes obsolete the following ones:
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6193d172e0887d08319a63e07720ac9f7f08a841/controllers/rosacluster_controller.go#L55
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6193d172e0887d08319a63e07720ac9f7f08a841/bootstrap/eks/controllers/eksconfig_controller.go#L62
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6193d172e0887d08319a63e07720ac9f7f08a841/controllers/awsmanagedcluster_controller.go#L54
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6193d172e0887d08319a63e07720ac9f7f08a841/controllers/rosacluster_controller.go#L55
...and many other lines.

Although we could optimize the role rules, I think these  are more intuitive for developers, since it's pretty straightforward trying to understand the objects controlled by the various controllers.

File: api/v1beta2/awscluster_webhook.go (hunk 10 above)
Comment (prometherion) [reply api/v1beta2/awscluster_webhook.go:302]: We still would need the check of each subfield, unfortunately. We could slender the webhook file, of course, but from a UX perspective, knowing which field is breaking the validation would be helpful.

File: api/v1beta2/awscluster_types.go [hunk 12]
```diff
@@ -229,7 +230,7 @@ type AWSLoadBalancerSpec struct {
 
 	// LoadBalancerType sets the type for a load balancer. The default type is classic.
```
Comment (vincepri) [inline api/v1beta2/awscluster_types.go:231]: Could we add a comment on what each value for this does and what the new value is useful for?

File: api/v1beta2/awscluster_webhook.go [hunk 13]
```diff
@@ -298,5 +298,49 @@ func (r *AWSCluster) validateControlPlaneLBs() field.ErrorList {
 		}
 	}
 
+	if r.Spec.ControlPlaneLoadBalancer.LoadBalancerType == LoadBalancerTypeDisabled {
```
Comment (vincepri) [inline api/v1beta2/awscluster_webhook.go:301]: Are we validating somewhere that the `type` cannot be changed once created? How are we dealing right now if a user creates the type = `disabled` and then wants `alb` and vice-versa?

File: api/v1beta2/awscluster_webhook.go [hunk 14]
```diff
@@ -298,5 +298,49 @@ func (r *AWSCluster) validateControlPlaneLBs() field.ErrorList {
...
+		if r.Spec.ControlPlaneLoadBalancer.DisableHostsRewrite {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "disableHostsRewrite"), r.Spec.ControlPlaneLoadBalancer.DisableHostsRewrite, "cannot disable hosts rewrite if the LoadBalancer reconciliation is disabled"))
+		}
+	}
+
+	for _, rule := range r.Spec.ControlPlaneLoadBalancer.IngressRules {
+		if (rule.CidrBlocks != nil || rule.IPv6CidrBlocks != nil) && (rule.SourceSecurityGroupIDs != nil || rule.SourceSecurityGroupRoles != nil) {
+			allErrs = append(allErrs, field.Invalid(field.NewPath("spec", "controlPlaneLoadBalancer", "ingressRules"), r.Spec.ControlPlaneLoadBalancer.IngressRules, "CIDR blocks and security group IDs or security group roles cannot be used together"))
+		}
+	}
```
Comment (vincepri) [inline api/v1beta2/awscluster_webhook.go:343]: We should open an issue to track for the next API version to rework this, it seems to me that we're going around the current API structure to disable the control plane management

File: controllers/awscluster_controller.go [hunk 15]
```diff
@@ -266,6 +266,45 @@ func (r *AWSClusterReconciler) reconcileDelete(ctx context.Context, clusterScope
 	return nil
 }
 
+func (r *AWSClusterReconciler) reconcileLoadBalancer(clusterScope *scope.ClusterScope, awsCluster *infrav1.AWSCluster) (*time.Duration, error) {
+	retryAfterDuration := 15 * time.Second
```
Comment (vincepri) [inline controllers/awscluster_controller.go:270]: Could we rework this to just return an `error` and leave the `retryAfterDuration` outside?

File: controllers/awscluster_controller.go [hunk 16]
```diff
@@ -447,3 +465,29 @@ func (r *AWSClusterReconciler) requeueAWSClusterForUnpausedCluster(_ context.Con
 		}
 	}
 }
+
+func (r *AWSClusterReconciler) checkForExternalControlPlaneLoadBalancer(clusterScope *scope.ClusterScope, awsCluster *infrav1.AWSCluster) *time.Duration {
+	requeueAfterPeriod := 15 * time.Second
+
```
Comment (vincepri) [inline controllers/awscluster_controller.go:471]: Same as above, the retry after should be outside in a caller

File: controllers/awsmachine_controller.go [hunk 17]
```diff
@@ -200,10 +202,16 @@ func (r *AWSMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 
 	infrav1.SetDefaults_AWSMachineSpec(&awsMachine.Spec)
 
+	cp, err := r.getControlPlane(ctx, log, cluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
```
Comment (vincepri) [inline controllers/awsmachine_controller.go:208]: It seems we're missing that the control plane is an optional reference, and this is going to panic when nil

File: controllers/awsmachine_controller.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (vincepri) [reply controllers/awsmachine_controller.go:1209]: Let's move these to a single place above

File: pkg/cloud/scope/machine.go [hunk 18]
```diff
@@ -371,6 +377,10 @@ func (m *MachineScope) IsEKSManaged() bool {
...
```
Comment (vincepri) [inline pkg/cloud/scope/machine.go:380]: A bit confusing on why this is needed, the control plane reference can be anything, including a new CRD; why do we need to mark it as externally managed?

File: controllers/awsmachine_controller.go (hunk 17 above)
Comment (prometherion) [reply controllers/awsmachine_controller.go:208]: Potentially could be solved with #4817.
